scene/gui/spin_box.cpp

```diff
@@ -137,25 +139,51 @@ void SpinBox::_release_mouse() {
 	}
 }
 
+void SpinBox::_mouse_exited() {
+	if (state_cache.up_button_hovered || state_cache.down_button_hovered) {
+		state_cache.up_button_hovered = false;
+		state_cache.down_button_hovered = false;
+		queue_redraw();
+	}
+}
+
 void SpinBox::gui_input(const Ref<InputEvent> &p_event) {
 	ERR_FAIL_COND(p_event.is_null());
 
 	if (!is_editable()) {
 		return;
 	}
 
+	Ref<InputEventMouse> me = p_event;
 	Ref<InputEventMouseButton> mb = p_event;
+	Ref<InputEventMouseMotion> mm = p_event;
 
 	double step = get_custom_arrow_step() != 0.0 ? get_custom_arrow_step() : get_step();
 
-	if (mb.is_valid() && mb->is_pressed()) {
-		bool up = mb->get_position().y < (get_size().height / 2);
+	Vector2 mpos;
+	bool mouse_on_up_button = false;
+	bool mouse_on_down_button = false;
+	if (mb.is_valid() || mm.is_valid()) {
+		Rect2 up_button_rc = Rect2(sizing_cache.buttons_left, sizing_cache.buttons_top, sizing_cache.buttons_width, sizing_cache.button_up_height);
```

Comment:
Why these are 4 separate values instead of caching it as Rect2?

Reply:
I would say this is a style detail and I have not seen any consensus on it so far. Some values are shared between both button rectangles and used in calculations, such as buttons_width.

Reply:
Then maybe you could use Vector2 instead? (for position and width) There is a Rect2 constructor from vectors.